src/merkle/tiered_smt/mod.rs

```diff
@@ -0,0 +1,211 @@
+use super::{BTreeMap, Felt, Rpo256, RpoDigest, Serializable, Vec, Word};
+
+mod bits;
+pub use bits::BitsIterator;
+
+mod content;
+pub use content::{Content, ContentType};
+
+mod index;
+pub use index::TreeIndex;
+
+mod storage;
+pub use storage::{Storage, StorageError};
+
+#[cfg(test)]
+mod tests;
+
+// TIERED SPARSE MERKLE TREE
+// ================================================================================================
+
+pub struct TieredSmt {
+    // TODO this should be a constant
+    empty_subtrees: Vec<RpoDigest>,
+    storage: Storage,
+}
+
+impl TieredSmt {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    pub const MAX_DEPTH: usize = 64;
+    pub const TIER_DEPTH: usize = 4;
+
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
```

Comment:
We need a constructor here to construct the SMT from entries (i.e., `(Word, Word)`).

Reply:
This looks straightforward for the current in-memory only structure, but that won't be the case when we use robust storage for large key/value pairs.

The current API allows us to create a tree only from a storage, and we might have such method for, let's say, an in-memory storage implementation. For something like rocksDB, we will have to provide more information such as the filesystem path for the root of the internal files.

The choice here was to delegate this question entirely to the storage. wdyt?

Reply:
I'm not yet sure what would be the right way to implement SMTs backed by different storage mechanisms. It might be that the best way to handle that would be to have a different struct (e.g., `RocksSmt`). At this point, I'd rather not worry about supporting multiple storage backends.

Let's get the basic things working (which I think will be complex enough), and then we get figure out how to move to support more complex storage mechanisms.

Reply:
Done! The storage is simplified into a small set of maps that will just index the required things for the algorithms to work.

In the future, we need to think on how to store the nodes. We have a couple of clever solutions around (i.e. jellyfish that optimizes to have ordered inserts) that we might get inspiration from.